lib/repl.js

```diff
@@ -1507,72 +1508,60 @@ function regexpEscape(s) {
+      return function(pos, message) {
+        nextToken.call(this, pos, message);
+
+        if (parser.type === acorn.tokTypes.eof) recoverable = true;
+      };
+    });
 
-    if (message === 'missing ) after argument list') {
-      const frames = e.stack.split(/\r?\n/);
-      const pos = frames.findIndex((f) => f.match(/^\s*\^+$/));
-      return pos > 0 && frames[pos - 1].length === frames[pos].length;
-    }
+    parser.extend('raise', (raise) => {
+      return function(pos, message) {
+        switch (message) {
+          case 'Unterminated template':
+          case 'Unterminated comment':
+            recoverable = true;
+            break;
+
+          case 'Unterminated string constant':
+            const token = parser.input.slice(parser.lastTokStart, parser.pos);
+            recoverable = token.endsWith('\\\n');
+        }
 
-    if (message === 'Invalid or unexpected token')
-      return isCodeRecoverable(code);
-  }
-  return false;
-}
+        raise.call(this, pos, message);
+      };
+    });
+  };
 
-// Check whether a code snippet should be forced to fail in the REPL.
-function isCodeRecoverable(code) {
-  var current, previous, stringLiteral;
-  var isBlockComment = false;
-  var isSingleComment = false;
-  var isRegExpLiteral = false;
-  var lastChar = code.charAt(code.length - 2);
-  var prevTokenChar = null;
-
-  for (var i = 0; i < code.length; i++) {
-    previous = current;
-    current = code[i];
-
-    if (previous === '\\' && (stringLiteral || isRegExpLiteral)) {
-      current = null;
-    } else if (stringLiteral) {
-      if (stringLiteral === current) {
-        stringLiteral = null;
-      }
-    } else if (isRegExpLiteral && current === '/') {
-      isRegExpLiteral = false;
-    } else if (isBlockComment && previous === '*' && current === '/') {
-      isBlockComment = false;
-    } else if (isSingleComment && current === '\n') {
-      isSingleComment = false;
-    } else if (!isBlockComment && !isRegExpLiteral && !isSingleComment) {
-      if (current === '/' && previous === '/') {
-        isSingleComment = true;
-      } else if (previous === '/') {
-        if (current === '*') {
-          isBlockComment = true;
-          // Distinguish between a division operator and the start of a regex
-          // by examining the non-whitespace character that precedes the /
-        } else if ([null, '(', '[', '{', '}', ';'].includes(prevTokenChar)) {
-          isRegExpLiteral = true;
-        }
-      } else {
-        if (current.trim()) prevTokenChar = current;
-        if (current === '\'' || current === '"') {
-          stringLiteral = current;
-        }
-      }
-    }
+  // For similar reasons as `defaultEval`, wrap expressions starting with a
+  // curly brace with parenthesis.  Note: only the open parenthesis is added
+  // here as the point is to test for potentially valid but incomplete
+  // expressions.
+  if (/^\s*\{/.test(code) && isRecoverableError(e, `(${code}`)) return true;
+
```

Comment:
Here I think it is very important that the regular expression exactly match the one in `defaultEval` as it would be problematic for this code to treat a given input as recoverable only to later find out that `defaultEval` would consider it to be a syntax error when done.